cmd/whitesourceExecuteScan.go

```diff
@@ -0,0 +1,247 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/SAP/jenkins-library/pkg/command"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"github.com/SAP/jenkins-library/pkg/versioning"
+	"github.com/SAP/jenkins-library/pkg/whitesource"
+)
+
+func whitesourceExecuteScan(config whitesourceExecuteScanOptions, telemetryData *telemetry.CustomData) {
+	// for command execution use Command
+	c := command.Command{}
+	sys := whitesource.NewSystem(config.ServiceURL, config.OrgToken, config.UserToken)
+	// reroute command output to logging framework
+	c.Stdout(log.Writer())
+	c.Stderr(log.Writer())
+
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	err := runWhitesourceScan(&config, sys, telemetryData, &c)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runWhitesourceScan(config *whitesourceExecuteScanOptions, sys whitesource.System, telemetryData *telemetry.CustomData, command *command.Command) error {
+	err := resolveProjectIdentifiers(command, config)
+	if err != nil {
+		return err
+	}
+
+	// Start the scan
+	projectsScanned, err := triggerWhitesourceScan(command, config, sys)
+	if err != nil {
+		return err
+	}
+	// Scan finished
+
+	log.Entry().Info("-----------------------------------------------------")
+	log.Entry().Infof("Project name: '%s'", config.ProjectName)
+	log.Entry().Infof("Product Version: '%s'", config.ProductVersion)
+	log.Entry().Infof("Project Token: %s", config.ProjectToken)
+	log.Entry().Infof("Number of projects scanned: %v", len(projectsScanned))
+	log.Entry().Info("-----------------------------------------------------")
+
+	if config.Reporting {
+		var links []piperutils.Path
+		for _, proj := range projectsScanned {
+			proj.Name = strings.Split(proj.Name, " - ")[0]
+			link, err := downloadRiskReport(proj.Token, proj.Name, sys)
+			if err != nil {
+				return err
+			}
+			links = append(links, *link)
+		}
+
+		// publish pdf file locations
+		piperutils.PersistReportsAndLinks("whitesourceExecuteScan", "./", nil, links)
+	}
+	return nil
+}
+
+func resolveProjectIdentifiers(command *command.Command, config *whitesourceExecuteScanOptions) error {
+	artifact, err := versioning.GetArtifact(config.ScanType, config.BuildDescriptorFile, &versioning.Options{}, command)
+	if err != nil {
+		return err
+	}
+	gav, err := artifact.GetCoordinates()
+	if err != nil {
+		return err
+	}
+	projectName, projectVersion := versioning.DetermineProjectCoordinates(config.ProjectName, config.DefaultVersioningModel, gav)
+	if config.ProjectName == "" {
+		config.ProjectName = projectName
+	}
+	if config.ProductVersion == "" {
+		config.ProductVersion = projectVersion
+	}
```

Comment:
Determining the coordinates can take significant time. For maven, it means firing up a JVM for each coordinate. On my pretty fast Mac, it can take several seconds locally.

```suggestion
	if config.ProjectName == "" || config.ProductVersion == "" {
		projectName, projectVersion := versioning.DetermineProjectCoordinates(config.ProjectName, config.DefaultVersioningModel, gav)
		if config.ProjectName == "" {
			config.ProjectName = projectName
		}
		if config.ProductVersion == "" {
			config.ProductVersion = projectVersion
		}
	}
```